drivers/ethernet/nxp_imx_netc/dsa_nxp_imx_netc.c

```diff
@@ -4,281 +4,148 @@
 
-#include <zephyr/device.h>
-#include <zephyr/drivers/gpio.h>
-#include <zephyr/kernel.h>
-#include <errno.h>
-#include <zephyr/sys/util.h>
 #include <zephyr/net/ethernet.h>
-#include <zephyr/linker/sections.h>
-#include <zephyr/toolchain/common.h>
 #include <zephyr/drivers/pinctrl.h>
-#include <zephyr/net/phy.h>
-#include <zephyr/devicetree.h>
 
+#include "../eth.h"
+#include "nxp_imx_netc.h"
 #include "fsl_netc_switch.h"
 
 #define DT_DRV_COMPAT nxp_netc_switch
-#include "nxp_imx_netc.h"
-
 #define PRV_DATA(ctx) ((struct dsa_netc_data *const)(ctx)->prv_data)
 
+struct dsa_netc_port_config {
+	const struct pinctrl_dev_config *pincfg;
+};
+
 struct dsa_netc_data {
-	int port_num;
-	int port_init_count;
 	swt_config_t swt_config;
 	swt_handle_t swt_handle;
-	const struct device *dev_master;
 	netc_cmd_bd_t *cmd_bd;
 };
 
-struct dsa_netc_slave_config {
-	/** MAC address for each LAN{123.,} ports */
-	uint8_t mac_addr[6];
-	const struct device *phy_dev;
-	netc_hw_mii_mode_t phy_mode;
-	volatile bool pseudo_mac;
-	const struct pinctrl_dev_config *pincfg;
-	int port_idx;
-	const struct device *ethernet_connection;
-};
-
-int dsa_netc_port_init(const struct device *dev)
+static int dsa_netc_port_init(const struct device *dev)
 {
-	const struct dsa_netc_slave_config *cfg = dev->config;
-	struct dsa_context *ctx = dev->data;
-	struct dsa_netc_data *prv = PRV_DATA(ctx);
+	const struct dsa_port_config *cfg = dev->config;
+	struct dsa_netc_port_config *prv_cfg = cfg->prv_config;
+	struct dsa_switch_context *dsa_switch_ctx = dev->data;
+	struct dsa_netc_data *prv = PRV_DATA(dsa_switch_ctx);
 	swt_config_t *swt_config = &prv->swt_config;
-
-	status_t result;
 	int err;
 
-	/* Get default config for whole switch before first port init */
-	if (prv->port_init_count == 0) {
-		SWT_GetDefaultConfig(swt_config);
-		swt_config->bridgeCfg.dVFCfg.portMembership = 0x0;
-	}
-
-	prv->port_init_count++;
-
-	if (prv->dev_master == NULL) {
-		prv->dev_master = cfg->ethernet_connection;
-	}
-
-	if (cfg->pseudo_mac) {
-		goto port_init;
+	if (prv_cfg->pincfg) {
+		err = pinctrl_apply_state(prv_cfg->pincfg, PINCTRL_STATE_DEFAULT);
+		if (err) {
+			return err;
+		}
 	}
 
-	err = pinctrl_apply_state(cfg->pincfg, PINCTRL_STATE_DEFAULT);
-	if (err) {
-		return err;
+	/* Get default config when beginning to init the first port */
+	if (dsa_switch_ctx->init_ports == 1) {
+		SWT_GetDefaultConfig(swt_config);
+		swt_config->bridgeCfg.dVFCfg.portMembership = 0x0;
 	}
 
-port_init:
 	/* miiSpeed and miiDuplex will get set correctly when link is up */
-	swt_config->ports[cfg->port_idx].ethMac.miiMode = cfg->phy_mode;
+	swt_config->ports[cfg->port_idx].ethMac.miiMode =
+		(strcmp(cfg->phy_mode, "mii") == 0
+			 ? kNETC_MiiMode
+			 : (strcmp(cfg->phy_mode, "rmii") == 0
+				    ? kNETC_RmiiMode
+				    : (strcmp(cfg->phy_mode, "rgmii") == 0
+					       ? kNETC_RgmiiMode
+					       : (strcmp(cfg->phy_mode, "gmii") == 0
+							  ? kNETC_GmiiMode
+							  : kNETC_RmiiMode))));
```

Comment:
use with `DT_INST_ENUM_HAS_VALUE()` in the `DSA_NETC_PORT_INST_INIT()`, so we don't have to compare strings at runtime.

Reply:
@maass-hamburg Thanks.
That's true, but using `DT_INST_ENUM_HAS_VALUE()` returns maybe a device specific value to use, like `kNETC_MiiMode` here.

Then it seems meaningless of the `const char *phy_mode;` definition in header to parsing node string property.
```
struct dsa_port_config {
        uint8_t mac_addr[6];
        const bool use_random_mac_addr;
        const int port_idx;
        const struct device *phy_dev;
        const char *phy_mode;
        const struct device *ethernet_connection;
        /** Instance specific config */
        void *prv_config;
};
```

As phy mode initialization is just once at beginning, I'm not sure if it affects a lot.
Or, should we remove `const char *phy_mode;` in header to let device driver to define?
Thanks.

Reply:
``const char *phy_mode`` would then have to use the same type as ``miiMode``.

Reply:
Thanks. Let me use `netc_hw_mii_mode_t phy_mode` in device driver to decide during build.